fdbserver/workloads/SpecialKeySpaceCorrectness.actor.cpp

```diff
@@ -212,6 +224,78 @@ struct SpecialKeySpaceCorrectnessWorkload : TestWorkload {
 		return GetRangeLimits(rowLimits, byteLimits);
 	}
 
+	ACTOR Future<Void> testCrossModuleRead(Database cx_, SpecialKeySpaceCorrectnessWorkload* self) {
+		Database cx = cx_->clone();
+		state Reference<ReadYourWritesTransaction> tx = Reference(new ReadYourWritesTransaction(cx));
+		// begin key outside module range
+		try {
+			wait(success(tx->getRange(
+			    KeyRangeRef(LiteralStringRef("\xff\xff/transactio"), LiteralStringRef("\xff\xff/transaction0")),
+			    CLIENT_KNOBS->TOO_MANY)));
+			ASSERT(false);
+		} catch (Error& e) {
+			if (e.code() == error_code_actor_cancelled) throw;
+			ASSERT(e.code() == error_code_special_keys_cross_module_read);
+			tx->reset();
+		}
+		// end key outside module range
+		try {
+			wait(success(tx->getRange(
+			    KeyRangeRef(LiteralStringRef("\xff\xff/transaction/"), LiteralStringRef("\xff\xff/transaction1")),
+			    CLIENT_KNOBS->TOO_MANY)));
+			ASSERT(false);
+		} catch (Error& e) {
+			if (e.code() == error_code_actor_cancelled) throw;
+			ASSERT(e.code() == error_code_special_keys_cross_module_read);
+			tx->reset();
+		}
+		// both begin and end outside module range
+		try {
+			wait(success(tx->getRange(
+			    KeyRangeRef(LiteralStringRef("\xff\xff/transaction"), LiteralStringRef("\xff\xff/transaction1")),
+			    CLIENT_KNOBS->TOO_MANY)));
+			ASSERT(false);
+		} catch (Error& e) {
+			if (e.code() == error_code_actor_cancelled) throw;
+			ASSERT(e.code() == error_code_special_keys_cross_module_read);
+			tx->reset();
+		}
+		// legal range read using the module range
+		try {
+			wait(success(tx->getRange(
+			    KeyRangeRef(LiteralStringRef("\xff\xff/transaction/"), LiteralStringRef("\xff\xff/transaction0")),
+			    CLIENT_KNOBS->TOO_MANY)));
+			TEST(true);
+			tx->reset();
+		} catch (Error& e) {
+			throw;
+		}
+		// begin keySelector outside module range
+		try {
+			const KeyRef key = LiteralStringRef("\xff\xff/cluster_file_path");
+			KeySelector begin = KeySelectorRef(key, false, 0);
+			KeySelector end = KeySelectorRef(keyAfter(key), false, 1);
+			wait(success(tx->getRange(begin, end, GetRangeLimits(CLIENT_KNOBS->TOO_MANY))));
+			ASSERT(false);
+		} catch (Error& e) {
+			if (e.code() == error_code_actor_cancelled) throw;
+			ASSERT(e.code() == error_code_special_keys_cross_module_read);
+			tx->reset();
+		}
+		// end keySelector inside module range, a tricky corner case
+		try {
+			tx->addReadConflictRange(singleKeyRange(LiteralStringRef("testKey")));
+			KeySelector begin = KeySelectorRef(readConflictRangeKeysRange.begin, false, 1);
+			KeySelector end = KeySelectorRef(LiteralStringRef("\xff\xff/transaction0"), false, 0);
+			wait(success(tx->getRange(begin, end, GetRangeLimits(CLIENT_KNOBS->TOO_MANY))));
+			TEST(true);
+			tx->reset();
+		} catch (Error& e) {
+			throw;
+		}
+		return Void();
```

Comment:
Can we add a test for no module found as well?

Reply:
Yes. I should do. Sorry for missing this one.

Reply:
Done. Thanks.